sensors/odometer.go

```diff
@@ -0,0 +1,183 @@
+						return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, replayTimestampErrorMessage)
+					}
+				}
+			}
+
+			if readingTimeOrientation == defaultTime || readingTimeOrientation.Sub(readingTimePosition).Milliseconds() < 0 {
+				ctxWithMetadata, md := contextutils.ContextWithMetadata(ctx)
+				if orientation, err = odom.Odometer.Orientation(ctxWithMetadata, make(map[string]interface{})); err != nil {
+					return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, "Orientation error")
+				}
+
+				readingTimeOrientation = time.Now().UTC()
+				if timeRequestedMetadata, ok := md[contextutils.TimeRequestedMetadataKey]; ok {
+					replay = true
+					if readingTimeOrientation, err = time.Parse(time.RFC3339Nano, timeRequestedMetadata[0]); err != nil {
+						return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, replayTimestampErrorMessage)
+					}
+				}
+			}
+			if math.Abs(float64(readingTimeOrientation.Sub(readingTimePosition).Milliseconds())) < replayTimeToleranceMsec {
+				return TimedOdometerSensorReadingResponse{
+					Position:    position,
+					Orientation: orientation,
+					ReadingTime: readingTimePosition.Add(readingTimePosition.Sub(readingTimeOrientation) / 2),
+					Replay:      replay,
+				}, nil
+			}
+		}
+	}
+}
+
+// NewOdometer returns a new Odometer.
+func NewOdometer(
+	ctx context.Context,
+	deps resource.Dependencies,
+	movementSensorName string,
+	dataFrequencyHz int,
+	logger golog.Logger,
+) (TimedOdometerSensor, error) {
+	_, span := trace.StartSpan(ctx, "viamcartographer::sensors::NewOdometer")
+	defer span.End()
+	if movementSensorName == "" {
+		return Odometer{}, nil
+	}
+	movementSensor, err := movementsensor.FromDependencies(deps, movementSensorName)
+	if err != nil {
+		return Odometer{}, errors.Wrapf(err, "error getting movement sensor \"%v\" for slam service", movementSensorName)
+	}
+
+	// A movement_sensor used as an Odometer must support Position and Orientation.
+	properties, err := movementSensor.Properties(ctx, make(map[string]interface{}))
+	if err != nil {
+		return Odometer{}, errors.Wrapf(err, "error getting movement sensor properties from \"%v\" for slam service", movementSensorName)
+	}
+	if !(properties.PositionSupported && properties.OrientationSupported) {
+		return Odometer{}, errors.New("configuring odometer movement sensor error: " +
+			"'movement_sensor' must support both Position and Orientation")
+	}
+
+	return Odometer{
+		name:            movementSensorName,
+		dataFrequencyHz: dataFrequencyHz,
+		Odometer:        movementSensor,
+	}, nil
+}
+
+// ValidateGetOdometerData checks every sensorValidationIntervalSec if the provided Odometer
+// returned valid timed readings every sensorValidationIntervalSec
+// until either success or sensorValidationMaxTimeoutSec has elapsed.
+// returns an error if at least one invalid reading was returned.
+func ValidateGetOdometerData(
```

Comment:
[opt] Why is this function needed? I would expect it to be sufficient to call GetProperties.

Reply:
This is true, GetProperties will already make initial calls to the cloud/actual hardware, therefore the only check we need is that the device returns GetProperties that shows the odom endpoints are supported

Reply:
This is the case only for replay sensors. In the case of live movement sensors, this check will still help us determine whether or not we're able to get the data we need.

Reply:
Note: Only localization on the robot is using the actual sensors at this point

Reply:
Should we remove the validation for both the Lidar and MovementSensor then?

I was also wondering if this level of validation is necessary and don't feel strongly about it.

Reply:
Though I do lean a bit towards removing the validation.